### PR TITLE
haskellPackages: Avoid re-enabling previously disabled tests

### PR DESCRIPTION
**Pull request by wolfgangwalther**
## Description of changes

The intent of all `doCheck = <condition>`, where condition is possibly true, is to disable the tests in a specific case. However, as currently written, this also has the effect of re-enabling the tests, even if they have been disabled by an override before, e.g. to `mkDerivation`.

This also affects the default value given in mkDerivation, which is `!isCross`. Before this change, aeson for example, would have been built with tests when cross-compiling, which was not intended.

The proper way is to set the `doCheck = false` attribute only conditionally, and otherwise rely on a previous override or the default value given in `mkDerivation`.

## Things done

The first commit fixes things as described.

The effect can be seen when trying to build `pkgsCross.musl64.haskellPackages.aeson`. This is currently broken, because the hadrian-based GHC 9.6.x doesn't provide `ghc-iserv`, so it can't build template haskell. But this fails when building the **tests**, which should have been disabled before via the default in mkDerivation: `doCheck ? !isCross`.

After this commit, `pkgsCross.musl64.haskellPackages.aeson` builds without tests and thus succeeds.

The second commit introduces a `dontCheckIf` utility function to avoid such problems in the future.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sternenseemann** [comment]
The alternative is of course to have the following which doesn't need a new library function.

```
overrideCabal (drv: {
  doCheck = !stdenv.hostPlatform.is32bit && drv.doCheck;
}
```

> This is currently broken, because the hadrian-based GHC 9.6.x doesn't provide ghc-iserv, so it can't build template haskell.

FTR, I don't think that's the issue. It would surprise me if GHC uses iserv somehow without `-fexternal-interpreter` being passed.

File: pkgs/development/haskell-modules/lib/compose.nix [hunk 1]
```diff
@@ -108,6 +108,13 @@ rec {
      of test suites listed in the package description file.
    */
   dontCheck = overrideCabal (drv: { doCheck = false; });
+  /* The dontCheckIf variant sets doCheck = false if the condition
+     applies. In any other case the previously set/default value is used.
+     This prevents accidentally re-enabling tests in a later override.
+   */
+  dontCheckIf = condition: overrideCabal (drv:
+    pkgs.lib.optionalAttrs condition { doCheck = false; }
+  );
```
Comment (maralorn) [inline pkgs/development/haskell-modules/lib/compose.nix:117]: Maybe you can simplify this to this?
```suggestion
  dontCheckIf = condition: if condition then dontCheck else lib.id;
```

**Review by maralorn — APPROVED** [review]
This looks reasonable to me.
I have one small suggestion, unless there is dissent I while merge this once we settled on a solution there.

**Comment by maralorn** [comment]
Great, thank you!

Really appreciate the high-quality contributions you made in the last weeks!

**Pull request merged**